src/runner.rs

```diff
@@ -166,53 +167,63 @@ enum RunResult {
 /// Function interpreter.
 pub struct Interpreter {
 	value_stack: ValueStack,
-	call_stack: Vec<FunctionContext>,
-	return_type: Option<ValueType>,
+	call_stack: StackWithLimit<FunctionContext>,
 	state: InterpreterState,
 }
 
 impl Interpreter {
-	pub fn new(func: &FuncRef, args: &[RuntimeValue]) -> Result<Interpreter, Trap> {
-		let mut value_stack = ValueStack::with_limit(DEFAULT_VALUE_STACK_LIMIT);
-		for &arg in args {
-			let arg = arg.into();
-			value_stack.push(arg).map_err(
-				// There is not enough space for pushing initial arguments.
-				// Weird, but bail out anyway.
-				|_| Trap::from(TrapKind::StackOverflow),
-			)?;
-		}
-
-		let mut call_stack = Vec::new();
-		let initial_frame = FunctionContext::new(func.clone());
-		call_stack.push(initial_frame);
-
-		let return_type = func.signature().return_type();
-
-		Ok(Interpreter {
-			value_stack,
+	/// Initialize an interpreter that will use `value_stack` and `call_stack`.
+	///
+	/// `value_stack` `call_stack` determine the allowed stack size during later executions.
+	pub fn new(
+		value_stack: StackWithLimit<RuntimeValueInternal>,
+		call_stack: StackWithLimit<FunctionContext>,
+	) -> Interpreter {
+		Interpreter {
+			value_stack: ValueStack(value_stack),
 			call_stack,
-			return_type,
 			state: InterpreterState::Initialized,
-		})
+		}
 	}
 
-	pub fn state(&self) -> &InterpreterState {
+	// Todo, use types to prevent user from calling start_execution on an inretpreter not in Initialized
+	// state.
+	/// Wipe all data in this interpreter, so it can be safely reused.
+	pub fn reset(&mut self) {
+		self.value_stack.truncate(0);
+		self.call_stack.truncate(0);
+		self.state = InterpreterState::Initialized;
+	}
+
+	pub(crate) fn state(&self) -> &InterpreterState {
 		&self.state
 	}
 
-	pub fn start_execution<'a, E: Externals + 'a>(
+	pub(crate) fn start_execution<E: Externals>(
```

Comment:
Why did you make this crate-internal?

Reply:
3. It doesn't type check function args.

Reply:
4. (writing these down for completeness) There is never a situation where `resume_execution` should be called before `start_execution`.